wip: add typescript

maxrimue opened the pull request:
Closes #5 

Oof.

I am creating this work in progress PR now to show what I've been working on, but this is far from complete. To sum it all up: Get TypeScript into the codebase, make tooling work with it and establish basic linting. Again, oof.

Here's a couple of learnings I wanted to write down:
- TypeScript's compiler (`tsc`) itself does quite a bit of linting. There's some stuff you can configure in `tsconfig.json`
- Unless configured otherwise, VS Code works pretty seamless with this already, errors are shown instantly and suggestions are made
- TypeScript (right now) complains whenever there's no typings for an external library, which (might?) suck. However, all libraries I've encountered so far, like React and React Native have types maintained for them. They can be installed via `npm i --save-dev @types/<library>` and are used automatically. I want to point out though that these might be community maintained and I'm not sure you can always be certain about their quality
- You will want to use `.ts` for any file now, since `strict` mode (which is enabled per configuration in `tsconfig.json`) does not like passive typing
- Use `.tsx` anytime you have XML-like elements in your code (what used to be JSX here), this should do the trick and not bother TypeScript any more than that
- I've used `concurrently` here to enable something like `startAndWatch`, it will run two things at the same time: Recompiling upon detected changes, and running expo. This should in theory allow for HMR to happen even with TypeScript, but we shall see
- Prettier doesn't yet check for `.ts` and `.tsx` files, but I'm keeping it that way until later to check for potential gotchas with TypeScript and Prettier getting into each other's way (you never know)
- There are _loads_ of errors we need to fix
- Like, really, over one hundred

There's still some stuff I'm uncertain about, such as whether we should also use TSLint, how we should configure linting, the tooling and scripting in the `package.json` and some other stuff. But anyway, this is only a WIP, and I've already come quite some way here I believe. I'd like to get this done sooner than later, considering that adoption will only become harder.

cykler commented:
First things first, nice job!

Just a couple of questions that came to my mind:
1. We don't need ESLint at all, right? Since TypeScript already has some rudimentary linting and TSLint will extend this even further, ESLint seems obsolete.
2. The thing about TypeScript complaining about external libraries worries me a bit though. What about small libraries that probably never will support TypeScript? Can we disable it in that case? 
I can imagine it's better having no types than no library.

Other than that, good job and let's get it up and running!

cykler commented on package.json:
```diff
@@ -1,11 +1,15 @@
+    "clean": "rimraf out",
+    "build": "npm run clean && tsc",
+    "buildAndWatch": "npm run build -- -w",
+    "start": "npm run build && concurrently \"npm run buildAndWatch\" \"npm run start\""
```
Wouldn't `npm run start` inside the `start` script lead to an infinite loop?

maxrimue replied:
Yup, fixed that 👍

maxrimue commented:
@cykler 

Indeed, ESLint probably won't be necessary. Even if we needed more linting, TSLint would be the better option as you already mentioned because of better compatibility I think. I agree with your point about necessary typings though. There seem to be two things you can do:
1. disable "noImplicitAny" as a linting rule (which causes the complaining), or
2. write a rudimentary "explanation" for TypeScript, which basically goes like
```
declare module "someModule" {
  export function something(lol: string): number; 
}
```
but I think it wouldn't even be necessary to type the whole module here, as long as you have an *explicit* any, you fix the problem TypeScript complains about.


I would like to continue the work of typing everything with you, as this is a great chance to refactor some stuff, learn about the codebase and figure out where to best put typings.

maxrimue commented:
@cykler In the last commit, `experiment with typing` (059c5ea), you can see about what I'm going for. The stuff I did there isn't yet verified to work, but syntax-wise it's working. Here are key-takeaways I'd like to share with you:
- when defining types, sometimes it does make sense to define them right in place, for example in the small helper functions I've touched, however, for bigger, and most importantly, _more often used combinations_, we need a central place for types
- this is also why I'd like to avoid general parameters like `state`. I think it's a pretty bad pattern being common in the React world, as you lose a sense of context and meaning, not to mention how easier it becomes to reason about
- there are comments in the form of `//* comment */` above the function declarations, these will appear in the same places where you also get type hints. For example, as you type a function's name in VS Code, it will display the comment as a string right next to the types. I like using it to give an extra sense of context

maxrimue commented on src/elements/Button.tsx:
```diff
@@ -1,8 +1,20 @@
 import theme from '../utils/theme';
 
-export default ({ children, highlight, style, ...props }) => {
+export interface Props extends TouchableOpacityProps {
```
If we're exporting this interface, we should probably give it a more descriptive name then right? Like `ButtonProps`? So that we do not pollute the global namespace. Btw, why do you export in this case? Are those typings required when using this component, or did you rather want to create a generic Prop type to be reused across components?

cykler replied:
Good point, these props are component specific. So exporting is not and renaming will not be necessary.

cykler commented:
The app does start again, but that's it. Nothing really works, but that is nothing we can't fix.

## Food for thoughts
* our whole TypeScript tooling feels very dirty and error-prone
* the expo dev tools don't start automatically when running `npm start`
* the app hot reloads too fast and therefore build changes will not be applied (so we basically have to save two times to see your changes take effect)

cykler commented:
@maxrimue FYI: I rebased, fixed merge conflicts and force pushed.

maxrimue commented:
I went over some online documentation now and improved a couple of things. Here's the list of improvements:
- no need to build before running the app
- no building and running at the same time necessary as well (therefore, also no `concurrently` anymore)
- instant live reload when editing/saving files
- both `npm start` and `expo start` work
- devtools now open automatically again upon starting

maxrimue commented:
I refactored the `persistedStore` code a bit to more align with how we work with actions and stores in other situations. This way, I think it's a bit easier to understand, but feel free to criticise. By refactoring, I also fixed a (few?) bug(s), the app seems to work normally again.

cykler commented:
@persistedStore: I like it! 

What do you think about merging this PR soon-ish? It' seems to work and is already hard enough to review. I think further refactoring should be handled in smaller increments.

maxrimue commented:
Thanks! Yeah I wash thinking the same thing - we should probably merge as soon as we don't "lose" any capability in master by merging, for example linting, app starting etc. But I think the PR wouldn't break anything in it's current state. Is there anything you think we should work on before merging? The typing should perhaps happen more ob the fly-ish and not all at once.

maxrimue commented on src/.expo/settings.json:
```diff
@@ -0,0 +1,7 @@
+  "lanType": "ip",
+  "dev": true,
+  "minify": false,
+  "urlRandomness": null
```
Is this change and the "packager-info.json" necessary @cykler?

cykler replied:
Don't know where this came from but I removed it for now.

maxrimue commented:
> ![anmerkung 2019-02-24 202922](https://user-images.githubusercontent.com/7421183/53304214-0fdc4880-3873-11e9-9865-5394453c81fb.jpg)
------------------------------------------
> ![anmerkung 2019-02-24 203024](https://user-images.githubusercontent.com/7421183/53304215-123ea280-3873-11e9-8cc7-5b18f03724a2.jpg)

# _Welcome to the next Level_ 😎

maxrimue commented:
I suggest we finish work on the design next, let's take planning this offline when we meet
